Userland/Libraries/LibGfx/Path.cpp

```diff
@@ -393,136 +393,227 @@ struct RoundTrip {
+            // distance and which are within distance ‘stroke-width’ of the point on the subpath at
+            // that position.
+            VERIFY(start - path_length < 1e-6f);
+            VERIFY(end - path_length < 1e-6f);
+            size_t line_index = start_index;
+            while (subpath[line_index + 1].get<1>() - start < 1e-6f)
+                line_index++;
+
+            Vector<NonnullRefPtr<Segment const>> out_path;
+            Vector<NonnullRefPtr<Segment const>> in_path;
+
+            {
+                // Add the first line
+                float t = start - subpath[line_index].get<1>();
+                FloatPoint A = subpath[line_index].get<0>();
+                FloatPoint B = subpath[line_index + 1].get<0>();
+                FloatVector2 AB = { B.x() - A.x(), B.y() - A.y() };
+                AB.normalize();
+                FloatVector2 AB_normal = { -AB.y(), AB.x() };
+                AB_normal.normalize();
+                AB_normal *= stroke_properties.stroke_width / 2;
+                FloatPoint out = A + AB * t + AB_normal;
+                FloatPoint in = A + AB * t - AB_normal;
+                out_path.append(adopt_ref(*new MoveSegment(out)));
+                switch (stroke_properties.stroke_linecap) {
+                case StrokeProperties::StrokeLinecap::Butt:
+                    in_path.append(adopt_ref(*new LineSegment(out)));
+                    break;
+                case StrokeProperties::StrokeLinecap::Round:
+                    in_path.append(adopt_ref(*new EllipticalArcSegment(out, A + AB * t, { stroke_properties.stroke_width / 2, stroke_properties.stroke_width / 2 }, atan2(AB.x(), AB.y()), 0, AK::Pi<float>, false, false)));
+                    break;
+                case StrokeProperties::StrokeLinecap::Square:
+                    in_path.append(adopt_ref(*new LineSegment(out)));
+                    in_path.append(adopt_ref(*new LineSegment(out - AB * stroke_properties.stroke_width / 2)));
+                    in_path.append(adopt_ref(*new LineSegment(in - AB * stroke_properties.stroke_width / 2)));
+                    break;
+                }
+                in_path.append(adopt_ref(*new LineSegment(in)));
+            }
 
-    Vector<ActiveRange, 128> active_ranges;
-    active_ranges.ensure_capacity(pen_vertices.size());
-    for (auto i = 0; i < pen_vertex_count; i++) {
-        active_ranges.unchecked_append({ angle_between(wrapping_index(pen_vertices, i - 1), pen_vertices[i]),
-            angle_between(pen_vertices[i], wrapping_index(pen_vertices, i + 1)) });
-    }
+            while (line_index + 1 < line_count and subpath[line_index + 1].get<1>() - end < 1e-6f) {
+                // Add offseted line to path
+
+                // FIXME: We need an algorithm for loop removal.
+                // There is one in
+                // Tiller, W.; Hanson, E.G. (1984). Offsets of Two-Dimensional Profiles. , 4(9), 0–46.
+                // doi:10.1109/mcg.1984.275995
```

Comment:
Why? And could you link to a PDF?

Reply:
Maybe I am mistaken, but I think you should get one of these problems (depending on Winding Rule)

![screenshot-2023-06-09_01:54:42](https://github.com/SerenityOS/serenity/assets/8083073/397da01c-79d4-4887-a42e-5fe625c3380e)
![screenshot-2023-06-09_02:01:13](https://github.com/SerenityOS/serenity/assets/8083073/306ff60b-1eae-4ea2-8bec-6fffbfcc43ae)

You either overpaint which is a problem when opacity is not 1 or you don't paint the corner. I do not know how common this situation is neither if it is a real problem or not.

Sadly I cannot link the PDF. The article mentioned is a paid one and I only got access to it through my university. But what you shouldn't do is go to scihub and search for the doi. Because you might find the article for free and that is bad.

Reply:
That's not how nonzero filling should work. If you look across the scanline (left to right), at the intersection, the winding orders should be: -1, -1, 1, 1. This is nonzero within the line and zero outside (and the fill will correctly deal with the loops, and does not overpaint).

Reply:
That is great to know. I will delete the de FIXME :) Thanks